### PR TITLE
Implement RegexTokenizer

### LINKED ISSUE #1670: Add a configurable RegexTokenizer
The regex pattern woudl be a parameter of the tokenizer.
The tokenizer would "search" the regex and emit one token per match.

### PR DESCRIPTION
**Pull request by Gearme**
Implementation of a RegexTokenizer for #1670.

To make this a decently flexible solution I implemented support for capture groups as well as simple find matches: The RegexTokenStream will check on construction if there are any capture groups (besides the entire match) and use either `captures_read_at` for capture support or `find_at` for simple matches. Tests and examples included.

Closes https://github.com/quickwit-oss/tantivy/issues/1670

### CODE REVIEW

**Comment by Gearme** [comment]
I found some edge-cases in regex patterns which need addressing, converting back to draft.

**Comment by fulmicoton** [comment]
@Gearme @mkleen any consensus on which PR we should progress on to land?
I think groups can be useful.
 
@PSeitz can you take over and make sure we end up merging something?

File: src/tokenizer/regex_tokenizer.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+impl<'a> RegexTokenStream<'a> {
+    fn new(text: &'a str, regex: &Regex) -> RegexTokenStream<'a> {
+        let capture_locations = if regex.captures_len() > 1 {
+            Some(regex.capture_locations())
+        } else {
+            None
+        };
+        RegexTokenStream {
+            text,
+            regex: regex.clone(),
```
Comment (PSeitz) [inline src/tokenizer/regex_tokenizer.rs:129]: This is called for every text. Arc may be preferable to `clone`

File: src/tokenizer/regex_tokenizer.rs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+                true
+            })
+    }
+}
+
+impl<'a> TokenStream for RegexTokenStream<'a> {
+    fn advance(&mut self) -> bool {
+        let tail_index = self.regex_match.map_or(0, |regex_match| regex_match.end());
+        if self.capture_locations.is_some() {
+            while !self.next_capture_group() {
```
Comment (PSeitz) [inline src/tokenizer/regex_tokenizer.rs:196]: `if let Some()` would be preferable to `is_some` and `unwrap`

File: src/tokenizer/regex_tokenizer.rs (hunk 1 above)
Comment (adamreichold) [reply src/tokenizer/regex_tokenizer.rs:129]: See also https://github.com/quickwit-oss/tantivy/pull/1759#discussion_r1061421507

**Comment by adamreichold** [comment]
@PSeitz @fulmicoton What do you think about merging #1759 first for the basic implementation and then rebasing this PR on top of it to add support for capture groups? This way, everyone gets kudos and we should end up with two easy to review PR.

**Comment by Gearme** [comment]
@adamreichold Sounds good to me!

**Comment by fulmicoton** [comment]
@adamreichold done sorry for the delay.

**Comment by PSeitz** [comment]
@Gearme The other PR is merged now

**Comment by fulmicoton** [comment]
What do we do with this PR? @Gearme @PSeitz

**Comment by Gearme** [comment]
I'm afraid I've been rather busy recently, I'll try and get to finishing and proofing this feature as soon as possible.

**Comment by mkleen** [comment]
> I found some edge-cases in regex patterns which need addressing, converting back to draft.

@Gearme Could you maybe add some of the edge cases to the tests? Then I could take over if it's ok for you.

**Comment by fulmicoton** [comment]
Closing this PR as it seems stalled.
